lib/srv/db/proxyserver.go

```diff
@@ -189,16 +199,49 @@ func (s *ProxyServer) connectToSite(ctx context.Context) (net.Conn, error) {
 	// Upgrade the connection so the client identity can be passed to the
 	// remote server during TLS handshake. On the remote side, the connection
 	// received from the reverse tunnel will be handled by tls.Server.
-	siteConn = tls.Client(siteConn, tlsConfig)
-	return siteConn, nil
+	serviceConn = tls.Client(serviceConn, tlsConfig)
+	return serviceConn, nil
+}
+
+// proxyToService starts proxying all traffic received from Postgres client
+// between this proxy and Teleport database service over reverse tunnel.
+func (s *ProxyServer) proxyToService(ctx context.Context, clientConn, serviceConn io.ReadWriteCloser) (retErr error) {
+	errCh := make(chan error, 2)
+	go func() {
+		defer s.log.Debug("Stop proxying from client to service.")
+		defer serviceConn.Close()
+		defer clientConn.Close()
+		_, err := io.Copy(serviceConn, clientConn)
+		errCh <- err
+	}()
+	go func() {
+		defer s.log.Debug("Stop proxying from service to client.")
+		defer serviceConn.Close()
```

Comment:
Hmm, I'm not sure this is the best strategy to close both client in both handlers. Can either of Closes fail on reentry to Close? 
Usually, there's a sentinel goroutine that detects when both sides are done and closes the clients. `sync.WaitGroup` could be used to ensure that both transport goroutines are gone before exiting which could benefit the tests as well.

Reply:
Not sure what's the issue with this approach? It also seems to be consistent with how it's done in other parts of Teleport codebase (for example https://github.com/gravitational/teleport/blob/v6.0.0-alpha.1/lib/client/client.go#L949-L963).

Reply:
The issue is that Close might be not reentrant although that does not seem to be the case here - it might be just returning an error on the 2nd entry.